subsys/net/lib/nrf_cloud/src/nrf_cloud_pgps.c

```diff
@@ -0,0 +1,1585 @@
+	 * have some free, if a previous attempt to replace expired
+	 * predictions failed (e.g., due to lack of LTE connection)
+	 */
+	for (pnum = npgps_num_free(); pnum < last; pnum++) {
+		block = npgps_pointer_to_block((uint8_t *)index.predictions[pnum]);
+		__ASSERT((block != -1), "unexpected ptr:%p for Prediction num:%d",
+			 index.predictions[pnum], pnum);
+		npgps_free_block(block);
+	}
+
+	/* move predictions we are keeping to the start */
+	for (i = last; i < index.header.prediction_count; i++) {
+		pnum = i - last;
+		index.predictions[pnum] = index.predictions[i];
+	}
+
+	/* set prediction pointers for 'last' in the newly empty
+	 * entries to NULL
+	 */
+	for (pnum = index.header.prediction_count - last; pnum <
+	      index.header.prediction_count; pnum++) {
+		index.predictions[pnum] = NULL;
+	}
+	npgps_print_blocks();
+
+	/* update index and header for new first stored prediction */
+	get_prediction_day_time(last, &index.start_sec,
+				&index.header.gps_day,
+				&index.header.gps_time_of_day);
+	LOG_DBG("updated index to gps_sec:%lld, day:%u, time:%u",
+		index.start_sec, index.header.gps_day,
+		index.header.gps_time_of_day);
+}
+
+int nrf_cloud_pgps_notify_prediction(void)
+{
+	/* when current prediction is ready, callback handler with
+	 * PGPS_EVT_AVAILABLE and pointer to prediction
+	 */
+	int err;
+	int pnum;
+	struct nrf_cloud_pgps_prediction *prediction;
+
+	LOG_DBG("num_predictions:%d, replacement threshold:%d, force fail start:%d"
+		" cur:%d, test interrupted:%d",
+		NUM_PREDICTIONS, REPLACEMENT_THRESHOLD, FORCE_FAIL,
+		fail_count, TEST_INTERRUPTED ? INTERRUPTED_PNUM : -1);
+
+	LOG_INF("Searching for prediction");
+	err = nrf_cloud_pgps_find_prediction(&prediction);
+	if ((err == -ELOADING) || (err == -EAPPROXIMATE)) {
+		pgps_need_assistance = true;
+		err = 0;
+	} else if (err < 0) {
+		if (!pgps_need_assistance) {
+			pgps_need_assistance = true;
+			err = nrf_cloud_pgps_request_all();
+			if (err) {
+				LOG_ERR("Error while requesting pgps set: %d", err);
+			}
+		} else {
+			err = 0;
+		}
+	} else if ((err >= 0) && (err < NUM_PREDICTIONS)) {
+		pnum = err;
+		err = 0;
+		LOG_INF("Found P-GPS prediction %d", pnum);
+		pgps_need_assistance = false;
+		/* the application should now send data to modem with
+		 * nrf_cloud_pgps_inject()
+		 */
+		if (handler) {
+			handler(PGPS_EVT_AVAILABLE, prediction);
+		}
+	}
+	return err;
+}
+
+int nrf_cloud_pgps_find_prediction(struct nrf_cloud_pgps_prediction **prediction)
+{
+	int64_t cur_gps_sec;
+	int64_t offset_sec;
+	int64_t start_sec = index.start_sec;
+	int64_t end_sec = index.end_sec;
+	uint16_t cur_gps_day;
+	uint32_t cur_gps_time_of_day;
+	uint16_t start_day = index.header.gps_day;
+	uint32_t start_time = index.header.gps_time_of_day;
+	uint16_t period_min = index.header.prediction_period_min;
+	uint16_t count = index.header.prediction_count;
+	int err;
+	int pnum;
+	bool margin = false;
+
+	if (prediction == NULL) {
```

Comment:
It's mentioned in the docs for 'nrf_cloud_pgps_init()` that the subsystem must be initialized before use, so checks that it is actually initialized could be added the functions in the public API

Reply:
Indeed... I usually do this.  Thanks.